crypto/src/hash/rpo/digest.rs

```diff
@@ -0,0 +1,112 @@
+use super::DIGEST_SIZE;
+use crate::{Digest, Felt, StarkField};
+use core::slice;
+use winterfell::{ByteReader, ByteWriter, Deserializable, DeserializationError, Serializable};
+
+// DIGEST TRAIT IMPLEMENTATIONS
+// ================================================================================================
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub struct RpoDigest256([Felt; DIGEST_SIZE]);
+
+impl RpoDigest256 {
+    pub fn new(value: [Felt; DIGEST_SIZE]) -> Self {
+        Self(value)
+    }
+
+    pub fn as_elements(&self) -> &[Felt] {
+        &self.0
+    }
+
+    pub fn digests_as_elements(digests: &[Self]) -> &[Felt] {
+        let p = digests.as_ptr();
+        let len = digests.len() * DIGEST_SIZE;
+        unsafe { slice::from_raw_parts(p as *const Felt, len) }
+    }
+}
+
+impl Digest for RpoDigest256 {
+    fn as_bytes(&self) -> [u8; 32] {
+        let mut result = [0; 32];
+
+        result[..8].copy_from_slice(&self.0[0].as_int().to_le_bytes());
+        result[8..16].copy_from_slice(&self.0[1].as_int().to_le_bytes());
+        result[16..24].copy_from_slice(&self.0[2].as_int().to_le_bytes());
+        result[24..].copy_from_slice(&self.0[3].as_int().to_le_bytes());
+
+        result
+    }
+}
+
+impl Default for RpoDigest256 {
+    fn default() -> Self {
+        RpoDigest256([Felt::default(); DIGEST_SIZE])
+    }
+}
+
+impl Serializable for RpoDigest256 {
+    fn write_into<W: ByteWriter>(&self, target: &mut W) {
+        target.write_u8_slice(&self.as_bytes());
+    }
+}
+
+impl Deserializable for RpoDigest256 {
+    fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
+        // TODO: check if the field elements are valid?
```

Comment:
I wonder now if this is really necessary here. `Felt::new` expects a `u64` and hence the only way this could fail is if the method `read_u64` fails to deserialize. Thus, there is no need to check if the field element is a field element as long as the deserialization succeeds. 
Let me know if I missed something.

Reply:
I think the question here is whether a digest which consists of valid `u64` values but which are a not valid field elements should be considered valid or not.

If we don't enforce this check, then it is possible for two or more 32-byte binary strings to map to the same digest when in field element form. If we do enforce this check, some 32-byte binary string will not be valid digests.

If we want to be conservative (which maybe we should), we should probably put an extra check here and make sure all `u64` values are actually valid field elements.

Reply:
I agree, we should be conservative in this regard. I guess the question now is: are the `u64` values reduced modulo the $p$ or not? If not, then we should raise a `Deserialization` error.
I will add these checks in `read_from`.